core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -543,62 +534,95 @@ public IndexResult index(Index index) throws IOException {
+                            currentNotFoundOrDeleted = versionValue.isDelete();
+                            useLuceneUpdateDocument = versionValue.isDelete() == false; // last operation is not delete
+                        }
+                        if (index.versionType().isVersionConflictForWrites(currentVersion, index.version(), currentNotFoundOrDeleted)) {
+                            earlyResultOnPreFlightError = Optional.of(new IndexResult(
+                                new VersionConflictEngineException(shardId, index, currentVersion, currentNotFoundOrDeleted),
+                                currentVersion
+                            ));
+                            versionForIndexing = Versions.NOT_FOUND;
+                            seqNoForIndexing = SequenceNumbersService.UNASSIGNED_SEQ_NO;
+                            performOperation = false;
+                        } else {
+                            versionForIndexing = index.versionType().updateVersion(currentVersion, index.version());
+                            seqNoForIndexing = seqNoService().generateSeqNo();
+                            performOperation = true;
+                        }
                     }
-                }
-                final long expectedVersion = index.version();
-                Optional<IndexResult> resultOnVersionConflict;
-                try {
-                    final boolean isVersionConflict = checkVersionConflict(index, currentVersion, expectedVersion, deleted);
-                    resultOnVersionConflict = isVersionConflict ? Optional.of(new IndexResult(currentVersion, index.seqNo(), false))
-                        : Optional.empty();
-                } catch (IllegalArgumentException | VersionConflictEngineException ex) {
-                    resultOnVersionConflict = Optional.of(new IndexResult(ex, currentVersion, index.seqNo()));
+                } else if (canOptimizeAddDocument && mayHaveBeenIndexedBefore(index) == false) {
+                    // no need to deal with out of order delivery - we never saw this one
+                    useLuceneUpdateDocument = false;
+                    currentNotFoundOrDeleted = true;
+                    assert index.version() == 1L : "can optimize on replicas but incoming version is [" + index.version() + "]";
+                    seqNoForIndexing = index.seqNo();
+                    versionForIndexing = index.version();
+                    performOperation = true;
+                } else {
+                    // drop out of order operations
+                    assert index.versionType().versionTypeForReplicationAndRecovery() == index.versionType() :
+                        "resolving out of order delivery based on versioning but version type isn't fit for it. got ["
+                            + index.versionType() + "]";
+                    final LuceneOpStatus luceneOpStatus = checkLuceneOpStatusBasedOnVersions(index);
```

Comment:
the semantics of this method are weird. I pass in an operation and it returns OLDER if the given operation has a higher version? I would have expected the opposite. This semantics also make the rest of this method hard to read since it has to negate the return values in 2/3 of the cases. I think you should flip it.